src/Adapter/Grid/Search/Factory/SearchCriteriaWithCategoryParentIdFilterFactory.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * 2007-2018 PrestaShop
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Grid\Search\Factory;
+
+use Category;
+use Configuration;
+use Context;
+use PrestaShop\PrestaShop\Core\Grid\Search\Factory\DecoratedSearchCriteriaFactory;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteria;
+use PrestaShop\PrestaShop\Core\Grid\Search\SearchCriteriaInterface;
+use Shop;
+use Tools;
+
+/**
+ * Class SearchCriteriaWithCategoryParentIdFilterFactory
+ */
+final class SearchCriteriaWithCategoryParentIdFilterFactory implements DecoratedSearchCriteriaFactory
```

Comment:
we shouldn't have to "hack" the request of the user, but if we do it we should do it before entering the controller.

Reply:
we could dispatch an event and a Hook in Controller Argument Resolver, wdyt?

Reply:
> about resolving the data of SearchCriteria/Filters instance

This is why I suggest to add an Hook and an Event here https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Controller/ArgumentResolver/SearchParametersResolver.php#L103

Reply:
> we should be able to extend them.

why? you are not unable to replace the existing one in the controller and Symfony controllers are not overridable ^^

Also creating your own using the interface is really easy, isn't it?

Reply:
so, what part should we change? :thinking: 

```php
 // we get $filters (SearhcCriteriaInterface) that are resolved in SearchParametersResolver
public function indexAction(CategoryFilters $filters) 
{
    // here we decorate resolved $filters with dynamic `id_category_parent` that is not saved with filters as it is not part of filtering form
    $searchCriteriaFactory = $this->get('prestashop.adapter.grid.search.factory.search_criteria_with_category_parent_id');
    $searchCriteria = $searchCriteriaFactory->createFrom($filters);

    // usual workflow below
    // ...
}
```

Reply:
For me, the inject $filters in action should already have the resolved value of id_category_parent.

So, you can dispatch an event (and a hook) in SearchParametersResolver, and use an event listener update the object before it was sent inside the action arguments.

So, it's not about changing this code, but moving it outside of the controller :)

Reply:
hmm, it makes sense. :+1: